chain/index/indexer.go

```diff
@@ -0,0 +1,380 @@
+package index
+
+import (
+	"context"
+	"database/sql"
+	"sync"
+
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/crypto"
+
+	"github.com/filecoin-project/lotus/chain/actors/builtin"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/types/ethtypes"
+	"github.com/filecoin-project/lotus/lib/sqlite"
+)
+
+var _ Indexer = (*SqliteIndexer)(nil)
+
+// IdToRobustAddrFunc is a function type that resolves an actor ID to a robust address
+type IdToRobustAddrFunc func(ctx context.Context, emitter abi.ActorID, ts *types.TipSet) (address.Address, bool)
+type recomputeTipSetStateFunc func(ctx context.Context, ts *types.TipSet) error
+
+type preparedStatements struct {
+	insertEthTxHashStmt                   *sql.Stmt
+	getNonRevertedMsgInfoStmt             *sql.Stmt
+	getMsgCidFromEthHashStmt              *sql.Stmt
+	insertTipsetMessageStmt               *sql.Stmt
+	updateTipsetToRevertedStmt            *sql.Stmt
+	hasTipsetStmt                         *sql.Stmt
+	updateTipsetToNonRevertedStmt         *sql.Stmt
+	removeTipsetsBeforeHeightStmt         *sql.Stmt
+	removeEthHashesOlderThanStmt          *sql.Stmt
+	updateTipsetsToRevertedFromHeightStmt *sql.Stmt
+	updateEventsToRevertedFromHeightStmt  *sql.Stmt
+	isTipsetMessageNonEmptyStmt           *sql.Stmt
+	getMinNonRevertedHeightStmt           *sql.Stmt
+	hasNonRevertedTipsetStmt              *sql.Stmt
+	updateEventsToRevertedStmt            *sql.Stmt
+	updateEventsToNonRevertedStmt         *sql.Stmt
+	getMsgIdForMsgCidAndTipsetStmt        *sql.Stmt
+	insertEventStmt                       *sql.Stmt
+	insertEventEntryStmt                  *sql.Stmt
+}
+
+type SqliteIndexer struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	db *sql.DB
+	cs ChainStore
+
+	idToRobustAddrFunc       IdToRobustAddrFunc
+	recomputeTipSetStateFunc recomputeTipSetStateFunc
+
+	stmts *preparedStatements
+
+	gcRetentionEpochs   int64
+	reconcileEmptyIndex bool
+	maxReconcileTipsets uint64
+
+	mu           sync.Mutex
+	updateSubs   map[uint64]*updateSub
+	subIdCounter uint64
+
+	closeLk sync.RWMutex
+	closed  bool
+}
+
+func NewSqliteIndexer(path string, cs ChainStore, gcRetentionEpochs int64, reconcileEmptyIndex bool,
+	maxReconcileTipsets uint64) (si *SqliteIndexer, err error) {
+
+	if gcRetentionEpochs != 0 && gcRetentionEpochs <= builtin.EpochsInDay {
```

Comment:
```suggestion
	if gcRetentionEpochs != 0 && gcRetentionEpochs < builtin.EpochsInDay {
```

`2880` should be fine, right? we divide by `builtin.EpochsInDay`  so should end up with `1` which is fine.